sway-error/src/handler.rs

```diff
@@ -21,26 +22,94 @@ struct HandlerInner {
 }
 
 impl Handler {
+    pub fn from_parts(errors: Vec<CompileError>, warnings: Vec<CompileWarning>) -> Self {
+        Self {
+            inner: RefCell::new(HandlerInner { errors, warnings }),
+        }
+    }
+
     /// Emit the error `err`.
     pub fn emit_err(&self, err: CompileError) -> ErrorEmitted {
         self.inner.borrow_mut().errors.push(err);
-        ErrorEmitted { _priv: () }
+        ErrorEmitted
     }
 
     /// Emit the warning `warn`.
     pub fn emit_warn(&self, warn: CompileWarning) {
         self.inner.borrow_mut().warnings.push(warn);
     }
 
+    pub fn has_error(&self) -> bool {
+        !self.inner.borrow().errors.is_empty()
+    }
+
+    pub fn has_warning(&self) -> bool {
+        !self.inner.borrow().warnings.is_empty()
+    }
+
     /// Extract all the errors from this handler.
     pub fn consume(self) -> (Vec<CompileError>, Vec<CompileWarning>) {
         let inner = self.inner.into_inner();
         (inner.errors, inner.warnings)
     }
+
+    pub fn append(&self, other: Handler) {
+        let (errors, warnings) = other.consume();
+        for warn in warnings {
+            self.emit_warn(warn);
+        }
+        for err in errors {
+            self.emit_err(err);
+        }
+    }
+
+    pub fn dedup(&self) {
+        let mut inner = self.inner.borrow_mut();
+        inner.errors = dedup_unsorted(inner.errors.clone());
+        inner.warnings = dedup_unsorted(inner.warnings.clone());
+    }
 }
 
 /// Proof that an error was emitted through a `Handler`.
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct ErrorEmitted {
-    _priv: (),
+pub struct ErrorEmitted;
```

Comment:
I figured it might have been the intent, but:

1. It doesn't stop you from creating instances at all
2. We actually do need the ability to make up some error emission to bail in the case of `ErrorRecovery` types

I would definitely want to improve our type resolution error recovery mechanism not to have to do this, but I think that's out of scope for this.

Reply:
TL;DR; Let's extend the `Handler` to clearly support different use cases that we can encounter in error reporting, rather then letting every piece of code freely instantiate `ErrorEmitted` instances. 

Ad 1) I am a bit confused. How could have the instances be created outside of the `Handler`? We had a private field that could not be set outside of the module, means `ErrorEmitted` instance could have been created only inside of the module, effectively only in the `Handler`. Or am I missing something?

Ad 2) I've seen this `ErrorRecovery` use case. Yet, I wouldn't solve it by allowing uncontrolled creating of `ErrorEmitted` instances. Here is why:
- with the new `Handler` and the way it is passed down we are getting a clear pattern for error reporting. If we have particular use cases in the error reporting, `Handler` should explicitly support it and offer a pattern for it. E.g. in this case something like `handler.emit_silent_err()` or something. (The same reasoning goes for `error_emitted` variable but I'll comment on that in the PR.)
- if we allow creation of `ErrorEmitted`, sooner or later they will start creeping through the code and be used for various workarounds bypassing the `Handler` and causing who knows what kind of side effects. Guaranteed :-)